### PR TITLE
Distribute 'keyof' union types

### LINKED ISSUE #23618: Cannot index into generic type if unions involved
**TypeScript Version:**  2.8.1

**Search Terms:** typescript intersection union "cannot be used to index"

**Code**

```ts
// This works:
type DBBoolTable<K extends string> = { [k in K]: 0 | 1 } 
enum Flag {
  FLAG_1 = "flag_1",
  FLAG_2 = "flag_2"
}

type SimpleDBRecord<Flag extends string> = { staticField: number } & DBBoolTable<Flag>
function getFlagsFromSimpleRecord<Flag extends string>(record: SimpleDBRecord<Flag>, flags: Flag[]) {
  // stub
  return record[flags[0]]
}

// This fails:

type DynamicDBRecord<Flag extends string> = ({ dynamicField: number } | { dynamicField: string }) & DBBoolTable<Flag>
function getFlagsFromDynamicRecord<Flag extends string>(record: DynamicDBRecord<Flag>, flags: Flag[]) {
  // stub
  return record[flags[0]] // Error: Type 'Flag' cannot be used to index type 'DynamicDBRecord<Flag>'
}
```

I know in this example I can just say `dynamicField: number | string`, but in my actual code there's a more complex type than just one field like that (and contains outside constraints like DB schema).

**Expected behavior:**
I should be able to index into the db record in the second example even with that dynamic field, because the DBRecord type says it should have all the flags in it.

**Actual behavior:**
It is unhappy with indexing into the type with a union in it, with the error mentioned above.

[**Playground Link**](https://www.typescriptlang.org/play/#src=type%20DBBoolTable%3CK%20extends%20string%3E%20%3D%20%7B%20%5Bk%20in%20K%5D%3A%200%20%7C%201%20%7D%20%0D%0Aenum%20Flag%20%7B%0D%0A%20%20FLAG_1%20%3D%20%22flag_1%22%2C%0D%0A%20%20FLAG_2%20%3D%20%22flag_2%22%0D%0A%7D%0D%0A%0D%0Atype%20SimpleDBRecord%3CFlag%20extends%20string%3E%20%3D%20%7B%20staticField%3A%20number%20%7D%20%26%20DBBoolTable%3CFlag%3E%0D%0Afunction%20getFlagsFromSimpleRecord%3CFlag%20extends%20string%3E(record%3A%20SimpleDBRecord%3CFlag%3E%2C%20flags%3A%20Flag%5B%5D)%20%7B%0D%0A%20%20return%20record%5Bflags%5B0%5D%5D%0D%0A%7D%0D%0A%0D%0Atype%20DynamicDBRecord%3CFlag%20extends%20string%3E%20%3D%20(%7B%20dynamicField%3A%20number%20%7D%20%7C%20%7B%20dynamicField%3A%20string%20%7D)%20%26%20DBBoolTable%3CFlag%3E%0D%0Afunction%20getFlagsFromDynamicRecord%3CFlag%20extends%20string%3E(record%3A%20DynamicDBRecord%3CFlag%3E%2C%20flags%3A%20Flag%5B%5D)%20%7B%0D%0A%20%20return%20record%5Bflags%5B0%5D%5D%0D%0A%7D)

**Related Issues:**
Not sure

### PR DESCRIPTION
**Pull request by ahejlsberg**
With this PR we transform `keyof` applied to a union type to an intersection of `keyof` applied to each union constituent. In other words, we rewrite types of the form `keyof (A | B)` to `keyof A & keyof B`.

Some examples:

```ts
type A = { a: string, c: string };
type B = { b: string, c: string };

type T1 = keyof (A | B);  // "c"
type T2<T> = keyof (T | B);  // (keyof T & "b") | (keyof T & "c")
type T3<U> = keyof (A | U);  // ("a" | keyof U) | ("c" & keyof U)
type T4<T, U> = keyof (T | U);  // keyof T & keyof U
type T5 = T2<A>;  // "c"
type T6 = T3<B>;  // "c"
type T7 = T4<A, B>;  // "c"
```

This PR is the mirror image of #22300 and finally brings symmetry to the type eqivalences:

```ts
keyof (A & B) <==> keyof A | keyof B
keyof (A | B) <==> keyof A & keyof B
```

Why it took so long to figure out I don't know!

Fixes #23618.

### CODE REVIEW

File: tests/baselines/reference/computedPropertyNames28_ES5.types [hunk 1]
```diff
@@ -12,8 +12,8 @@ class C extends Base {
 >super : typeof Base
 
         var obj = {
->obj : { [x: string]: () => void; }
->{            [(super(), "prop")]() { }        } : { [x: string]: () => void; }
+>obj : { [(super(), "prop")](): void; }
```
Comment (weswigham) [inline tests/baselines/reference/computedPropertyNames28_ES5.types]: So, I don't know if it's related to this PR, but this change caught my eye:
```ts
{ [(super(), "prop")](): void; }
```
isn't a valid type. Somewhere we're reusing the name's expression tree when building the computed property name type and shouldn't be. :(

Comment (ahejlsberg) [reply tests/baselines/reference/computedPropertyNames28_ES5.types]: This is actually a baseline change from #23592. I somehow messed up a merge and ended up with a lot of apparent changes in this PR.

Anyways, what's not valid about the type? It's a method with a computed property name that uses the comma operator.

**Comment by ahejlsberg** [comment]
I somehow ended up with a bad merge on this branch so I'm abandoning this PR in favor of #23645.

File: tests/baselines/reference/computedPropertyNames28_ES5.types (hunk 1 above)
Comment (weswigham) [reply tests/baselines/reference/computedPropertyNames28_ES5.types]: That's supposed to be type output - if someone wrote `type Foo = { [(super(), "prop")](): void; }`, we'd issue an error.

**Pull request closed without merging**